src/Microsoft.ML.TimeSeries/SrCnnEntireAnomalyDetector.cs

```diff
@@ -0,0 +1,846 @@
+                    1.6994687082728674e-05, 9.767061541798089e-06
+                };
+
+            private readonly double _threshold;
+            private readonly double _sensitivity;
+            private readonly SrCnnDetectMode _detectMode;
+
+            //used in all modes
+            private double[] _predictArray;
+            private double[] _backAddArray;
+            private double[] _fftRe;
+            private double[] _fftIm;
+            private double[] _magList;
+            private double[] _magLogList;
+            private double[] _spectralList;
+            private double[] _transRe;
+            private double[] _transIm;
+            private double[] _ifftRe;
+            private double[] _ifftIm;
+            private double[] _ifftMagList;
+            private double[] _cumSumList;
+            private double[] _cumSumShift;
+            //used in AnomalyAndMargin mode
+            private double[] _units;
+            private double[] _val;
+            private double[] _ans;
+            private double[] _curWindow;
+
+            public SrCnnEntireModeler(double threshold, double sensitivity, SrCnnDetectMode detectMode)
+            {
+                _threshold = threshold;
+                _sensitivity = sensitivity;
+                _detectMode = detectMode;
+            }
+
+            public void Train(double[] values, ref double[][] results)
+            {
+                if (results == null)
+                {
+                    results = new double[values.Length][];
+                    for (int i = 0; i < results.Length; ++i)
+                    {
+                        results[i] = new double[_outputLengthArray[(int)_detectMode]];
+                    }
+                }
+                else if (results.Length > values.Length)
+                {
+                    Array.Resize<double[]>(ref results, values.Length);
+                }
+                SpectralResidual(values, results, _threshold);
+                //Optional Steps
+                if (_detectMode == SrCnnDetectMode.AnomalyAndMargin)
+                {
+                    GetMargin(values, results, _sensitivity);
+                }
+                else if (_detectMode == SrCnnDetectMode.AnomalyAndExpectedValue)
+                {
+                    GetExpectedValue(values, results);
+                }
+            }
+
+            private void AllocateDoubleArray(ref double[] arr, int length)
+            {
+                if (arr == null)
+                {
+                    arr = new double[length];
+                }
+                else if (arr.Length != length)
+                {
+                    Array.Resize<double>(ref arr, length);
+                }
+            }
+
+            private void SpectralResidual(double[] values, double[][] results, double threshold)
+            {
+                // Step 1: Get backadd wave
+                double[] backAddList = BackAdd(values);
```

Comment:
>BackAdd [](start = 39, length = 7)

I think this puts the values in `_backAddArray`, it doesn't need to return it.

Reply:
Fixed this and other similar places.

---
In reply to: [427856187](https://github.com/dotnet/machinelearning/pull/5135#discussion_r427856187) [](ancestors = 427856187)